docs/source/examples/md/sec_agg_example.md

```diff
@@ -0,0 +1,195 @@
+```
+
+In fl_server.py, we load and run the FL server job.  
+
+```
+server = FLServer()
+server_id = 0
+job_path = "fl_job_config"
+job = FLRunTimeJob()
+job.load_server_job(job_path, server_id)
+server.set_server_job(job)
+server.start()
+```
+
+In fl_trainer.py, we prepare the MNIST dataset, load and run the FL trainer job, then evaluate the accuracy.  Before training , we first prepare the party's private key and other party's public key. Then, each party generates mask using Diffie-Hellman key aggregate protocal with its parivate key and other's public key [1], and masks the model parameters before uploading to the server. Finally, the server can remove the masks by aggregating  the parameters from all the parties.
+
+```python
+train_reader = paddle.batch(
+    paddle.reader.shuffle(paddle.dataset.mnist.train(), buf_size=500),
+    batch_size=16)
+
+trainer_num = 2
+trainer_id = int(sys.argv[1]) # trainer id for each guest
+
+job_path = "fl_job_config"
+job = FLRunTimeJob()
+job.load_trainer_job(job_path, trainer_id)
+trainer = FLTrainerFactory().create_fl_trainer(job)
+trainer.start()
+
+output_folder = "fl_model"
+epoch_id = 0
+step_i = 0
+while not trainer.stop():
+    epoch_id += 1
+    print("epoch %d start train" % (epoch_id))
+    starttime = datetime.datetime.now()
+
+    # prepare the aggregated parameters
+    param_name_list = []
+    param_name_list.append("fc_0.b_0.opti.trainer_" + str(trainer_id))
+    param_name_list.append("fc_0.b_0.opti.trainer_" + str(trainer_id))
+
+    inputs = fluid.layers.data(name='x', shape=[1, 28, 28], dtype='float32')
+    label = fluid.layers.data(name='y', shape=[1], dtype='int64')
+    feeder = fluid.DataFeeder(feed_list=[inputs, label], place=fluid.CPUPlace())
+
+    scale = pow(10.0, 5)
+     # 1. load priv key and other's pub key
+    # party_name = Party(trainer_id + 1)
+    dh = DiffieHellman(group=15, key_length=256)
+    dh.load_private_key(str(trainer_id) + "_priv_key.txt")
+
+    digestmod="SHA256"
+
+    for data in train_reader():
+        step_i += 1
+        noise = 0.0
+
+        # 2. generate noise
+        secagg_starttime = datetime.datetime.now()
+        key = str(step_i).encode("utf-8")
+        for i in range(trainer_num):
+            if i != trainer_id:
+                f = open(str(i) + "_pub_key.txt", "r")
+                public_key = int(f.read())
+                dh.generate_shared_secret(public_key, echo_return_key=True)
+                msg = dh.shared_key.encode("utf-8")
+                hex_res1 = hmac.new(key=key, msg=msg, digestmod=digestmod).hexdigest()
+                current_noise = int(hex_res1[0:8], 16) / scale
+                if i > trainer_id:
+                    noise = noise + current_noise
+                else:
+                    noise = noise - current_noise
+        if step_i % 100 == 0:
+            print("Step: {0}".format(step_i))
+        # 3. add noise between training and sending.
+        accuracy, = trainer.run(feed=feeder.feed(data),
```

Comment:
I think there is a random mask here, please correct me if I am wrong.
``` python
trainer = FLTrainerFactory().create_fl_trainer(job)
```
Generally, there can be a lot of trainers here, I think param_name_list can be set in job information. And a secure aggregate trainer can be created through FLTrainerFactory.

``` python
for i in range(trainer_num):
            if i != trainer_id:
                f = open(str(i) + "_pub_key.txt", "r")
                public_key = int(f.read())
                dh.generate_shared_secret(public_key, echo_return_key=True)
                msg = dh.shared_key.encode("utf-8")
                hex_res1 = hmac.new(key=key, msg=msg, digestmod=digestmod).hexdigest()
                current_noise = int(hex_res1[0:8], 16) / scale
                if i > trainer_id:
                    noise = noise + current_noise
                else:
                    noise = noise - current_noise
```
During each step, a random noise can be generated in run(feed=feeder.feed(data),  fetch=["top_k_0.tmp_0"])
I don't think there should be param_name_list and mask arguments in run interface.

Reply:
Thanks for you kind comments. I have optimize that. "param_name_list" will be set in job information and noise/mask will be generate in the function "run".